File: persistence.py
# HaiveControl — LAN remote control & screen sharing with an AI/MCP interface.
# Copyright (C) 2026 The HaiveControl Authors.
# SPDX-License-Identifier: AGPL-3.0-or-later

import os
import shlex
import subprocess
import sys

APP_NAME = "HaiveControl"
LABEL = "com.haive.agent"


def boot_command(agent_args):
    """The command that relaunches this agent at boot, as a list of strings.
    Uses the frozen exe path when built, else `python server.py`."""
    if getattr(sys, "frozen", False):
        return [sys.executable] + agent_args
    return [sys.executable, os.path.abspath(sys.argv[0])] + agent_args


def install(agent_args):
    cmd = boot_command(agent_args)
    if sys.platform.startswith("win"):
        _install_windows(cmd)
    elif sys.platform == "darwin":
        _install_macos(cmd)
    else:
        _install_linux(cmd)


def uninstall():
    try:
        if sys.platform.startswith("win"):
            _uninstall_windows()
        elif sys.platform == "darwin":
            _uninstall_macos()
        else:
            _uninstall_linux()
    except Exception:
        pass


# ---- Windows: HKCU \...\Run ----

def _win_run_key(access):
    import winreg

    return winreg.OpenKey(
        winreg.HKEY_CURRENT_USER,
        r"Software\Microsoft\Windows\CurrentVersion\Run",
        0,
        access,
    )


def _install_windows(cmd):
    import winreg

    with _win_run_key(winreg.KEY_SET_VALUE) as key:
        winreg.SetValueEx(
            key, APP_NAME, 0, winreg.REG_SZ, subprocess.list2cmdline(cmd)
        )


def _uninstall_windows():
    import winreg

    with _win_run_key(winreg.KEY_SET_VALUE) as key:
        winreg.DeleteValue(key, APP_NAME)


# ---- macOS: LaunchAgent ----

def _plist_path():
    return os.path.expanduser(f"~/Library/LaunchAgents/{LABEL}.plist")


def _install_macos(cmd):
    args = "".join(f"\n      <string>{c}</string>" for c in cmd)
    plist = f"""<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0"><dict>
  <key>Label</key><string>{LABEL}</string>
  <key>ProgramArguments</key><array>{args}
  </array>
  <key>RunAtLoad</key><true/>
  <key>KeepAlive</key><true/>
</dict></plist>
"""
    path = _plist_path()
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        f.write(plist)
    subprocess.run(["launchctl", "load", path], check=False)


def _uninstall_macos():
    path = _plist_path()
    subprocess.run(["launchctl", "unload", path], check=False)
    if os.path.exists(path):
        os.remove(path)


# ---- Linux: XDG autostart (.desktop) ----

def _desktop_path():
    return os.path.expanduser(f"~/.config/autostart/{APP_NAME.lower()}.desktop")


def _install_linux(cmd):
    exec_line = " ".join(shlex.quote(c) for c in cmd)
    entry = (
        "[Desktop Entry]\n"
        "Type=Application\n"
        f"Name={APP_NAME}\n"
        f"Exec={exec_line}\n"
        "X-GNOME-Autostart-enabled=true\n"
    )
    path = _desktop_path()
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        f.write(entry)


def _uninstall_linux():
    path = _desktop_path()
    if os.path.exists(path):
        os.remove(path)

File: crates/agent/src/persistence.rs
// Standard, visible autostart per OS. Nothing hidden; uninstall removes it.
use std::env;
#[allow(unused_imports)]
use std::path::{Path, PathBuf};

pub fn install(args: &[String]) {
    let exe = env::current_exe().unwrap_or_default();
    #[cfg(windows)]
    win_install(&exe, args);
    #[cfg(target_os = "macos")]
    mac_install(&exe, args);
    #[cfg(all(unix, not(target_os = "macos")))]
    linux_install(&exe, args);
    let _ = (&exe, args);
}

/// Boot/logon-level autostart — a Scheduled Task (Windows), LaunchDaemon (macOS)
/// or systemd system service (Linux). More robust than `install` (per-user Run
/// key / LaunchAgent): survives reboots and restarts the agent if it dies.
/// Requires elevation to create; run the enrollment command as admin/root.
pub fn install_service(args: &[String]) {
    let exe = env::current_exe().unwrap_or_default();
    #[cfg(windows)]
    win_install_service(&exe, args);
    #[cfg(target_os = "macos")]
    mac_install_service(&exe, args);
    #[cfg(all(unix, not(target_os = "macos")))]
    linux_install_service(&exe, args);
    let _ = (&exe, args);
}

pub fn uninstall() {
    #[cfg(windows)]
    {
        win_uninstall();
        win_service_uninstall();
    }
    #[cfg(target_os = "macos")]
    {
        mac_uninstall();
        mac_service_uninstall();
    }
    #[cfg(all(unix, not(target_os = "macos")))]
    {
        linux_uninstall();
        linux_service_uninstall();
    }
}

fn home() -> String {
    env::var("HOME")
        .or_else(|_| env::var("USERPROFILE"))
        .unwrap_or_default()
}

// ---- macOS: LaunchAgent ----
#[cfg(target_os = "macos")]
fn plist_path() -> PathBuf {
    PathBuf::from(home()).join("Library/LaunchAgents/com.haive.agent.plist")
}

#[cfg(target_os = "macos")]
fn mac_install(exe: &Path, args: &[String]) {
    let mut pa = format!("      <string>{}</string>\n", exe.display());
    for a in args {
        pa.push_str(&format!("      <string>{a}</string>\n"));
    }
    let plist = format!(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<plist version=\"1.0\"><dict>\n  <key>Label</key><string>com.haive.agent</string>\n  <key>ProgramArguments</key><array>\n{pa}  </array>\n  <key>RunAtLoad</key><true/>\n  <key>KeepAlive</key><true/>\n</dict></plist>\n"
    );
    let p = plist_path();
    if let Some(dir) = p.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    let _ = std::fs::write(&p, plist);
    let _ = std::process::Command::new("launchctl")
        .args(["load", &p.to_string_lossy()])
        .status();
}

#[cfg(target_os = "macos")]
fn mac_uninstall() {
    let p = plist_path();
    let _ = std::process::Command::new("launchctl")
        .args(["unload", &p.to_string_lossy()])
        .status();
    let _ = std::fs::remove_file(&p);
}

// ---- Linux: XDG autostart ----
#[cfg(all(unix, not(target_os = "macos")))]
fn desktop_path() -> PathBuf {
    PathBuf::from(home()).join(".config/autostart/haivecontrol.desktop")
}

#[cfg(all(unix, not(target_os = "macos")))]
fn linux_install(exe: &Path, args: &[String]) {
    let mut ex = format!("{}", exe.display());
    for a in args {
        ex.push(' ');
        ex.push_str(a);
    }
    let entry = format!(
        "[Desktop Entry]\nType=Application\nName=HaiveControl\nExec={ex}\nX-GNOME-Autostart-enabled=true\n"
    );
    let p = desktop_path();
    if let Some(dir) = p.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    let _ = std::fs::write(&p, entry);
}

#[cfg(all(unix, not(target_os = "macos")))]
fn linux_uninstall() {
    let _ = std::fs::remove_file(desktop_path());
}

// ---- Windows: HKCU Run ----
#[cfg(windows)]
fn win_install(exe: &Path, args: &[String]) {
    use winreg::enums::*;
    use winreg::RegKey;
    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
    if let Ok(key) = hkcu.open_subkey_with_flags(
        r"Software\Microsoft\Windows\CurrentVersion\Run",
        KEY_SET_VALUE,
    ) {
        let mut cmd = format!("\"{}\"", exe.display());
        for a in args {
            cmd.push_str(&format!(" \"{a}\""));
        }
        let _ = key.set_value("HaiveControl", &cmd);
    }
}

#[cfg(windows)]
fn win_uninstall() {
    use winreg::enums::*;
    use winreg::RegKey;
    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
    if let Ok(key) = hkcu.open_subkey_with_flags(
        r"Software\Microsoft\Windows\CurrentVersion\Run",
        KEY_SET_VALUE,
    ) {
        let _ = key.delete_value("HaiveControl");
    }
}

// ---- Windows: Scheduled Task (starts at logon, elevated) ----
#[cfg(windows)]
fn win_install_service(exe: &Path, args: &[String]) {
    let mut tr = format!("\"{}\"", exe.display());
    for a in args {
        tr.push(' ');
        tr.push_str(a);
    }
    let _ = std::process::Command::new("schtasks")
        .args(["/Create", "/TN", "HaiveControl", "/TR", &tr, "/SC", "ONLOGON", "/RL", "HIGHEST", "/F"])
        .status();
}

#[cfg(windows)]
fn win_service_uninstall() {
    let _ = std::process::Command::new("schtasks")
        .args(["/Delete", "/TN", "HaiveControl", "/F"])
        .status();
}

// ---- macOS: LaunchDaemon (starts at boot, root) ----
#[cfg(target_os = "macos")]
fn daemon_path() -> PathBuf {
    PathBuf::from("/Library/LaunchDaemons/com.haive.agent.plist")
}

#[cfg(target_os = "macos")]
fn mac_install_service(exe: &Path, args: &[String]) {
    let mut pa = format!("      <string>{}</string>\n", exe.display());
    for a in args {
        pa.push_str(&format!("      <string>{a}</string>\n"));
    }
    let plist = format!(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<plist version=\"1.0\"><dict>\n  <key>Label</key><string>com.haive.agent</string>\n  <key>ProgramArguments</key><array>\n{pa}  </array>\n  <key>RunAtLoad</key><true/>\n  <key>KeepAlive</key><true/>\n</dict></plist>\n"
    );
    let p = daemon_path();
    let _ = std::fs::write(&p, plist);
    let _ = std::process::Command::new("launchctl").args(["load", &p.to_string_lossy()]).status();
}

#[cfg(target_os = "macos")]
fn mac_service_uninstall() {
    let p = daemon_path();
    let _ = std::process::Command::new("launchctl").args(["unload", &p.to_string_lossy()]).status();
    let _ = std::fs::remove_file(&p);
}

// ---- Linux: systemd system service (starts at boot, root) ----
#[cfg(all(unix, not(target_os = "macos")))]
fn unit_path() -> PathBuf {
    PathBuf::from("/etc/systemd/system/haivecontrol.service")
}

#[cfg(all(unix, not(target_os = "macos")))]
fn linux_install_service(exe: &Path, args: &[String]) {
    let mut ex = format!("{}", exe.display());
    for a in args {
        ex.push(' ');
        ex.push_str(a);
    }
    let unit = format!(
        "[Unit]\nDescription=HaiveControl agent\nAfter=network.target\n\n[Service]\nExecStart={ex}\nRestart=always\nRestartSec=5\n\n[Install]\nWantedBy=multi-user.target\n"
    );
    let _ = std::fs::write(unit_path(), unit);
    let _ = std::process::Command::new("systemctl").arg("daemon-reload").status();
    let _ = std::process::Command::new("systemctl").args(["enable", "--now", "haivecontrol.service"]).status();
}

#[cfg(all(unix, not(target_os = "macos")))]
fn linux_service_uninstall() {
    let _ = std::process::Command::new("systemctl").args(["disable", "--now", "haivecontrol.service"]).status();
    let _ = std::fs::remove_file(unit_path());
}
